[indexer-v2 1/n] add schemas and typesadd schemas and types

longbowlu opened the pull request:
## Description 

Breaking [this](https://github.com/MystenLabs/sui/pull/13574) into smaller pieces.
This is first PR to define schemas and basic data models. Addressed most comments in #13574 


Stack:
1. https://github.com/MystenLabs/sui/pull/13760
2. https://github.com/MystenLabs/sui/pull/13774
3. https://github.com/MystenLabs/sui/pull/13775
4. https://github.com/MystenLabs/sui/pull/13778

## Test Plan 

Testing is done in a live indexer with the big PR.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
[indexer-v2 1/n] add schemas and types

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-044020_events/up.sql:
```diff
@@ -0,0 +1,24 @@
+    transaction_digest          bytea        NOT NULL,
+    checkpoint_sequence_number  bigint       NOT NULL,
+    -- array of SuiAddress in bytes. All signers of the transaction.
+    senders                     bytea[]      NOT NULL,
```
Note for us on GraphQL that we need to reflect the fact that transactions can have multiple senders in `Event` and `TransactionBlock` in the GraphQL schema.

longbowlu replied:
that's why it's `bytea[]` or am i misunderstanding what you meant?

amnn replied:
This is a comment for the folks working on GraphQL that we need to update our schema to make use of the fact that this is an array (Currently GraphQL's `TransactionBlock` only has a scalar `sender: SuiAddress` field, not `senders: SuiAddress`).

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-060729_packages/up.sql:
Should we get rid of this? The table that we need for packages will look quite different, so it seems better to start it from scratch -- it's pretty trivial to extract the name and modules for a package from the package's BCS bytes (in the `objects` table).

longbowlu replied:
Can we leave this later? I want to have the basic skeleton up to work so we can have multiple people work on different areas simultaneously.

longbowlu replied:
more explicitly, I'm less confident in my ability to shortly figure out the final form of package table such that it supports package upgrade. And hence don't want us to be blocked here

amnn replied:
I'm all for getting a basic skeleton up -- but does it need to include a `packages` table at all? My point is that I think whatever table we end up having for packages will look completely different from this, so it would just involve deleting it later, which diminishes its value as a skeleton for me.

bmwill replied:
yeah i'm of the same opinion, lets drop this table and if we need to load any pkgs from the db for doing type resolution we can just read the object from the objects table

longbowlu replied:
yes it's needed as a ModuleResolver, which is next PR in the stack https://github.com/MystenLabs/sui/pull/13775/files. 
Reading it from the objects table surely works too, but with different code. Since that code won't live long too, I'd prefer to living with this until we figure out the package upgrade story.

longbowlu replied:
another thing part of package upgrade is when framework packages upgrade they don't obtain a new package ID, hence it's a bit tricky how to track them.

amnn replied:
> another part of package upgrade is when framework packages upgrade they don't obtain a new package ID, hence it's a bit tricky how to track them.

Sure, this part is the sell for me for keeping this table around (but I would still just store the package's BCS bytes, rather than explode it out into pairs of module name and module bytes).

longbowlu replied:
That means we always need to read and parse the entire package, but for now (before we integrate with the read queries) the only pattern we use is to get a particular module i.e. `get_module_by_id`. I think we all agree this table would look very different soon, so IMO it's not the part we should try to make it perfect (or even better) now.

amnn replied:
I don't think `get_module_by_id` is any more efficient with the current implementation -- it's a sequential scan looking for the name in either case, and should probably have an in-memory LRU cache in front of it, in the indexer and/or GraphQL service.

To be honest, the thing that bugs me about this table is (I think) similar to @bmwill's concern, and it's centered around this `bcs_bytes` type.  It feels quite special that it gets a definition, but then it's confusing because:

- It's got a generic name that suggests it holds any kind of bcs_bytes.
- But, then when you think about it, a generic type to store bcs bytes is just `bytea`, so why does it exist?
- Then when you look at the definition, you realise it has a "name" field, so it's not a generic type after all.
- Now you're trying to work out why you would need a special type for pairs of names and bcs bytes, and again you're at a loss because it doesn't make anything more efficient: The BCS representation of this type may even be byte-for-byte identical after the type is created, and it needs to be searched in the same way regardless.

So now I'm annoyed (I mean, not really, but pretend that I am), because I'm 80% through this schema, and I can't figure out why it exists, whereas if I had read this:

```sql
-- Index packages separately so framework packages don't get pruned.
CREATE TABLE packages 
(
    package_id                   bytea    PRIMARY KEY,
    modules                        bytea    NOT NULL
);
```

all would be right in the world.

longbowlu replied:
I actually agree with the argument that this schema is not efficient either :). The only reason it's like this is - i copied it from the v1 schema and it **works**. I didn't put too much thoughts into it because it's worth a separate task as we know. The "special" types are from the old code too (so perhaps @gegaowp  could explain why it looks like this?)

But now you've made it clear enough this piece bugs you, I wouldn't insist on keeping being lazy and would change it to your suggestion so we can move forward
```
CREATE TABLE packages 
(
    package_id                   bytea    PRIMARY KEY,
    modules                      bytea    NOT NULL
);
```

longbowlu replied:
and now it's updated

bmwill replied:
Glad we all agree on this!

gegaowp replied:
agreed too that the BCS array is not any more efficient to search for a specific module compared with storing BCS bytes directly.
this was introduced to support `get_normalized_move_modules_by_package` and `get_normalized_move_module` on indexer, but these 2 were never implemented on this table, but indeed if we want to avoid parsing the whole package each time, it's better to explode each package into rows of modules in the package.

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-044023_objects/up.sql:
```diff
@@ -0,0 +1,33 @@
+    object_id                   bytea         PRIMARY KEY,
+    object_version              bigint        NOT NULL,
+    object_digest               bytea         NOT NULL,
+    checkpoint_sequence_number  bigint        NOT NULL,
```
Objects don't seem to have an edge connecting them to the transaction that created them (i.e. `previousTransaction`).

longbowlu replied:
`previousTransaction` is part of `Object` which is bcs serialized in another field. When we query a tx we get its updated objects from `transaction` table. Are you asking this because client may query `previous_transaction` and only it?

amnn replied:
Got it -- yes, that's one of the fields that can be independently queried.  It's also one of the `ObjectFilter`-s.

Generally, having to deserialize the BCS to get access to some fields is not the end of the world.  For the second point, how easy would it be to run this query by querying the transaction table for the previous transaction, and then get the IDs of its changed objects?  If that seems like a reasonable strategy, then we can avoid the extra column.

bmwill replied:
That would just require another round-trip through the db, not the end of the world and we can optimize things as we go as well

longbowlu replied:
> how easy would it be to run this query by querying the transaction table for the previous transaction, and then get the IDs of its changed objects?

transaction table is designed to support this query, there's a `object_changes` field there. something like
```
select object_changes from transactions where transaction_digest = '';
```

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-044026_transactions/up.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE TABLE transactions (
+    tx_sequence_number          BIGINT       PRIMARY KEY,
```
Is `tx_sequence_number` alone the primary key? I thought it was local to a particular checkpoint.

longbowlu replied:
this is a global index that is (partially) used for pagination

amnn replied:
Can you always identify the transactions in a checkpoint by a range of sequence numbers?  If so, then this is probably fine.  Otherwise, this will complicate implementing consistency across all queries (like trying to consistently paginate transactions sent by an address, which needs to be restricted to the transactions at or before the latest checkpoint indexed at the time the first page was queried).

In GraphQL, cursors will be opaque, and will probably be represented by Base64 encoding compound keys, like `base64("c:<checkpoint_seq>;t:<tx_seq>;e:<ev_seq>")` for an event cursor.

bmwill replied:
Yes this number is intended to be the total order sequence of the transactions, and given a checkpoint you should be able to concretely tell which range of sequence numbers apply to it

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-044035_tx_indices/up.sql:
```diff
@@ -0,0 +1,33 @@
+CREATE INDEX tx_indices_input_objects ON tx_indices USING GIN(input_objects);
+CREATE INDEX tx_indices_changed_objects ON tx_indices USING GIN(changed_objects);
+CREATE INDEX tx_indices_senders ON tx_indices USING GIN(senders);
+CREATE INDEX tx_indices_recipients ON tx_indices USING GIN(recipients);
```
I think we talked about not including an index on the other kinds of address to transaction relationships, but in hindsight, it seems like it would be pretty impossible to use them in a query without such an index -- should we add one for payers?

longbowlu replied:
if that's a required feature then we have to create an index for it. Just to caveat that each index will add more burden to the ingestion pipeline, so we'd better be self-disciplined in adding them

amnn replied:
Would it be feasible to add it and then gate the code that handles post-processing so if a particular deployment doesn't want to deal with indexing address-transaction relationships, it can turn that bit off? We will have a similar functionality on the GraphQL side to disable the relevant aspects of the schema.

longbowlu replied:
We can check existence of table indices and create if needed when the instance starts. So it's doable, just not by diesel.

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-044044_checkpoints/up.sql:
```diff
@@ -0,0 +1,26 @@
+    network_total_transactions          bigint       NOT NULL,
+    previous_checkpoint_digest          bytea,
+    -- if this checkpoitn is the last checkpoint of an epoch
+    end_of_epoch                        boolean      NOT NULL,
```
In the GraphQL schema we put the end of epoch data on the last checkpoint in the epoch.  IIRC I did it this way because that's how it shows up on the protocol side, but in the data layer it looks like the end of epoch data resides on the epoch when it has ended.

Is that right?  If so, I'm tempted to get rid of `EndOfEpochData` on the checkpoint, and keep it just on `Epoch`.  If we do that, I don't think we need the `end_of_epoch` boolean either, because we can figure that out from the checkpoint's epoch instead (i.e. keep track of the epoch's first and last checkpoint instead of tracking whether each checkpoint is the last checkpoint of an epoch).

longbowlu replied:
>  IIRC I did it this way because that's how it shows up on the protocol side

you are right, this is what we are talking about in slack too. We are re-designing the data layer together so whatever makes the most sense will be in the new data layer. No constraint on this end.

> I'm tempted to get rid of EndOfEpochData on the checkpoint
I don't think it's worthwhile changing core structure if it's what you mean. Again on the data layer we can present it in the most desired way.

> I don't think we need the end_of_epoch boolean either, because we can figure that out from the checkpoint's epoch instead (i.e. keep track of the epoch's first and last checkpoint instead of tracking whether each checkpoint is the last checkpoint of an epoch).

When client queries a checkpoint, without this `end_of_epoch` boolean, you need to join it with `epoch` table. A boolean field is small enough that IMO is not worth the join

longbowlu replied:
`end_of_epoch_data` is part of existing rpc `Checkpoint` [struct](https://github.com/MystenLabs/sui/blob/main/crates/sui-json-rpc-types/src/sui_checkpoint.rs#L55). Is it ok to break it by always returning None? Or do we want to keep this column for now and delete it when json rpc is deprecated?

amnn replied:
I don't think we want to break the existing RPC, but we don't have to right?  Even if we changed the data model, the JSON-RPC endpoint could still construct an `EndOfEpochData` by querying its epoch and its next epoch, which is not very efficient, but lets us build the new data model while still supporting the old query pattern.

I'm happy with either approach (keep `EndOfEpochData` in the data layer, and shim it in GraphQL until JSON-RPC is deprecated, or remove `EndOfEpochData` in the data layer, and shim the access in JSON-RPC).

longbowlu replied:
> lets us build the new data model while still supporting the old query pattern.

I wouldn't worry too much about having a couple of more columns now. It's very easy to drop them when they are no long needed. But I don't feel strongly about this one either.

longbowlu replied:
and now it's updated

amnn reviewed (commented):
Mostly LGTM, with a few questions here and there!  Also, doesn't need to be tackled in this PR, but noting some of our schema related discussions from various places:

- Flattening the "end of epoch data" into the epoch table.
- Dedicated table for indexing "Display"
- (Related to epochs) Indexing system state summary (for protocol configs, system state version, system parameters, etc).  This can be accessed from the `objects` and eventual object history tables, but that means it will be subject to more aggressive pruning than is necessary given that we only produce one of these per epoch.  Being able to write queries such as to calculate historical APY have come up as a user request a couple of times.

gegaowp commented on crates/sui-indexer/src/types_v2.rs:
```diff
@@ -0,0 +1,473 @@
+    pub non_refundable_storage_fee: u64,
+    pub checkpoint_commitments: Vec<u8>,
+    pub validator_signature: Vec<u8>,
+    pub successful_tx_num: usize,
```
successful_tx_num is indexed but not persisted in the DB, why is that?

longbowlu replied:
The plan is to this to calculate real time tps, and that will be streaming `successful_tx_num` per checkpoint without reading it from db

gegaowp reviewed (commented):
overall looks great!

amnn commented on crates/sui-indexer/migrations_v2/2023-08-19-044035_tx_indices/up.sql:
```diff
@@ -0,0 +1,33 @@
+    packages                    bytea[]      NOT NULL,
+    -- array of "package::module" of all MoveCalls of the transaction.
+    -- e.g. "0x0000000000000000000000000000000000000000000000000000000000000003::sui_system"
+    package_modules             text[]       NOT NULL,
```
Good question, I think we did impose limits on identifier lengths.  Currently it is [128](https://github.com/MystenLabs/sui/blob/f56fa815d2d5db6b0d034de64ee1f543d13ba972/crates/sui-protocol-config/src/lib.rs#L1354).

oxade replied:
@gegaowp do you plan to impose this limit in the indexer too?

oxade replied:
@longbowlu

longbowlu replied:
the underlying storage is the same so for me no difference: https://www.postgresql.org/docs/current/datatype-character.html

gegaowp replied:
yes psql handles this well for us, so we do not have to

longbowlu commented on crates/sui-indexer/migrations_v2/2023-08-19-044020_events/up.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE TABLE events
+(
+    tx_sequence_number          BIGINT       NOT NULL,
```
yes it unique identifies a tx and is globally consistent

longbowlu commented on crates/sui-indexer/migrations_v2/2023-08-19-044026_transactions/up.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE INDEX transactions_checkpoint_sequence_number ON transactions (checkpoint_sequence_number);
+
+-- only create index for system transactions (0). See types_v2.rs
+CREATE INDEX transactions_transaction_kind ON transactions (transaction_kind) WHERE transaction_kind <> 0;
```
good catch! it should be `= 0`, will fix

longbowlu commented:
Update:
1. get rid of `end_of_epoch_data` in epoch table and only keep `epoch_commitments`
2. flatten `end_of_epoch_info` in epoch table into various field
3. still keep `end_of_epoch_data` in checkpoint table, but this can change depending on [this Q](https://github.com/MystenLabs/sui/pull/13760#discussion_r1326348296)

longbowlu commented:
A small inconvenience coming from flattening `EndOfEpochInfo` is: the protocol_version and reference_gas_price there are for **next** epoch. but querying the next epoch's row simply for these two values are a bit annoying. So I'd add two columns `next_epoch_reference_gas_price` and `next_epoch_protocol_version` in `epoch` tables instead.

amnn commented:
> A small inconvenience coming from flattening EndOfEpochInfo is: the protocol_version and reference_gas_price there are for next epoch. but querying the next epoch's row simply for these two values are a bit annoying. So I'd add two columns next_epoch_reference_gas_price and next_epoch_protocol_version in epoch tables instead.

@longbowlu, IIUC, this means the only way to calculate an epoch's RGP is to query the previous epoch for the next epoch RGP (same for protocol version).  Can we just change the semantics of this field instead -- i.e. that the `epochs` table stores the fields for the current Epoch, and the indexer populates the next epoch's table entry based on the previous epoch's `EndOfEpochData`?

longbowlu commented:
> > A small inconvenience coming from flattening EndOfEpochInfo is: the protocol_version and reference_gas_price there are for next epoch. but querying the next epoch's row simply for these two values are a bit annoying. So I'd add two columns next_epoch_reference_gas_price and next_epoch_protocol_version in epoch tables instead.
> 
> @longbowlu, IIUC, this means the only way to calculate an epoch's RGP is to query the previous epoch for the next epoch RGP (same for protocol version). Can we just change the semantics of this field instead -- i.e. that the `epochs` table stores the fields for the current Epoch, and the indexer populates the next epoch's table entry based on the previous epoch's `EndOfEpochData`?

No, we will have both `rgp` and `next_epoch_rgp` for one epoch. It's similar to [this discussion](https://github.com/MystenLabs/sui/pull/13760#discussion_r1326780197), which I guess demonstrates my appetite for "denormalization" when the data is small enough #12229

amnn commented:
> No, we will have both rgp and next_epoch_rgp for one epoch.

Ah okay, as long as these fields are accessible on their current epoch then I don't mind if it's denormalized -- although I do struggle to come up with a reason to access them on the previous epoch, as in, it just doesn't seem very discoverable.  If somebody asked me to write a query to get the rgp for the epoch after `E`, I would intuitively write:

```graphql
query { epoch(E + 1) { referenceGasPrice } }
```

Rather than:

```graphql
query { epoch(E) { nextReferenceGasPrice } }
```

Also, at the moment the `epochs` table only has a single `protocol_version` and a single `reference_gas_price` field, so I assumed these were from flattening `EndOfEpochInfo`, is that not the case, i.e. is the change to introduce the `next_*` variants still to be done, or did I miss something?

bmwill commented:
So one reason to keep around the "next_epoch_*" values is that they technically can change during the epoch and it may be useful to query this information and present it in various places. in a similar vain we have validators who have both data current for this epoch and data that i can change in preparation for the next epoch. But maybe that's something that's more clunky to query for in the api? especially since it can change at an arbitrary txn boundary.

longbowlu commented:
> although I do struggle to come up with a reason to access them on the previous epoch

+1. I have it there purely because jsonrpc `get_epoch` returns `EndOfEpochInfo` today, which contains these two next epoch things. In fact i'd be happy to nuke them, e.g. by always returning None, if no one depends on this field. IIRC @bmwill mentioned his confusion around whether these are "next_epoch" or current "epoch". Turns out it's not only confusing but also a bit controversial to deal with lol. (@gegaowp in case you know the context behind it).

@amnn the next epoch things are in next PR as mentioned: https://github.com/MystenLabs/sui/pull/13774/files#r1327518202

>  But maybe that's something that's more clunky to query for in the api? especially since it can change at an arbitrary txn boundary.
@bmwill yeah it's indeed awkward to expose those dynamic data in this table. I think a more reliable way to track that if one is interested is events.

gegaowp commented:
> a reason to access them on the previous epoch

agreed from data discoverability pov, tbh I do not exactly the contexts, it might be b/c the next_epoch_* data are calculated in the last checkpoint of the previous epoch rather than the first one of next epoch if I would guess.

longbowlu commented:
>  it might be b/c the next_epoch_* data are calculated in the last checkpoint of the previous epoch rather than the first one of next epoch if I would guess.

notably these are all part of the SuiSystemStateEvent (or sth like it) that is emitted after advancing the epoch.

gegaowp reviewed (approved):
looks great, thanks for answering / addressing all the comments!
![ship](https://media.giphy.com/media/bZBmitwUwKtDa/giphy-downsized-large.gif)

longbowlu commented:
I think we addressed most of the comments in this PR for now. The outstanding one is next_epoch_*, which is added in next PR https://github.com/MystenLabs/sui/pull/13774. Landing this for now.

amnn commented:
> So one reason to keep around the "next_epoch_*" values is that they technically can change during the epoch and it may be useful to query this information and present it in various places.

This is an interesting point, but is that how this data is currently filled though?  I thought it came (as @gegaowp mentioned) from the data in the last checkpoint in the epoch and so it's only available when the next epoch is about to start and when these numbers won't change, same for the event emitted by `advance_epoch`.

The dynamic version of this value would indeed be interesting, but also not straightforward to implement (have to scan for transactions that update the validator set for rgp-- and I don't think we have a good way to dynamically detect the next protocol version).

bmwill commented:
It would actually not be too difficult to implement, you'd just load the current version of the system state object and print out the existing info from there

amnn commented:
Is it directly a field on the system state object or would we have to calculate it based on all validators' proposed gas prices?  If it's just a field then yes it does indeed seem straightforward, but if it's the latter case then the complications for me are:

- having to reimplement the logic for calculating rgp from the proposals (not the end of the world -- and we do have to do this for some other stuff -- but it does bug me slightly).
- balancing calculating rgp on every request without caching (i.e. load the active validator set every time) -- less efficient but simpler -- with detecting when a checkpoint contains a PTB with a transaction that might affect the RGP (fiddly to get right, especially if the relevant functions in sui-system are `public`).

None of this stuff is insurmountable and the feature does seem useful, but it also sounds like it's quite fundamentally different from the semantics of the current `next_epoch_reference_gas_price`.  Enough that I would consider omitting it in the initial pass through.

longbowlu commented:
> Is it directly a field on the system state object or would we have to calculate it based on all validators' proposed gas prices?

In move, the calculation is only done at the epoch advancement moment, so it's not a field in 0x5 or its children.

>  and I don't think we have a good way to dynamically detect the next protocol version).

right, the voting info is not recorded on chain.

I don't disagree these could be useful but it seems very very niche (i.e. only a small subset of people care about it). So in order to implement it, we could consider handling them in a separate handler instead of the main one.

amnn commented:
@longbowlu, I would tend to agree -- does that mean it's okay to omit the `next_*` fields?
